segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/MultiKeyReaderWriterScheduler.java

```diff
@@ -0,0 +1,188 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.metadata;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.val;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A scheduler utility that implements pattern similar to Multiple Readers - Single Writer pattern.
+ */
+public class MultiKeyReaderWriterScheduler {
```

Comment:
This is a very complex class with a lot of things that can go wrong, so it will be a likely candidate for tasks to "hang indefinitely". Is there a reason why:
1. This is needed?
2. A reader-writer lock is needed? Can `MultiKeyAsyncSequentialProcessor` be used instead (it also serializes based on arbitrary string keys).

Reply:
This data structure is used to control concurrent access to the SLTS metadata cache. We need to be able to have multiple reader/single writer access to the metadata in order to avoid deadlocks that we saw during 0.8 testing.
The use case is bit different